web/assets/js/actions.ts

```diff
@@ -1,214 +1,391 @@
+            saveToLocalStorage(defaultMnemonicKey, mnemonic)
+        }
+
+        // Initialize the wallet using the saved mnemonic
+        this.wallet = await GnoWallet.fromMnemonic(mnemonic);
+
+        // Initialize the provider
+        this.provider = new GnoWSProvider(wsURL)
+
+        // Connect the wallet to the provider
+        this.wallet.connect(this.provider)
+
+        // Faucet token initialization //
+        let faucetToken: string | null = localStorage.getItem(defaultFaucetTokenKey)
+        if (faucetToken && faucetToken !== "") {
+            // Faucet token initialized
+            this.faucetToken = faucetToken
+        }
+    }
+
+    /**
+     * Saves the faucet token to local storage
+     * @param token the faucet token
+     */
+    public setFaucetToken(token: string) {
+        this.faucetToken = token
+
+        localStorage.setItem(defaultFaucetTokenKey, token)
+    }
+
+    /**
+     * Fetches the saved faucet token, if any
+     */
+    public getFaucetToken(): string | null {
+        return this.faucetToken
+    }
+
+    /****************
+     * GAME ENGINE
+     ****************/
+
+    /**
+     * Joins the waiting lobby for the game
+     * @param time
+     */
+    public async joinLobby(time: GameTime): Promise<GameSettings> {
+        // Join the waiting lobby
+        const joinResponse: BroadcastTxCommitResult = await this.wallet?.callMethod(
+            chessRealm,
+            "JoinLobby", // TODO change when API is ready
+            [
+                time.time.toString(),
+                time.increment.toString()
+            ],
+            TransactionEndpoint.BROADCAST_TX_COMMIT
+        ) as BroadcastTxCommitResult
+
+        // Parse the response from the node
+        const joinDataRaw: string | null = joinResponse.deliver_tx.ResponseBase.Data
+        if (!joinDataRaw) {
+            throw new Error("invalid join lobby response")
+        }
+
+        // Magically parse the response
+        const joinData: GamePromise = JSON.parse(joinDataRaw)
+
+        // Wait to be matched with an opponent
+        return await this.waitForGame(joinData.id)
+    }
+
+    /**
+     * Waits for the game with the specified ID to begin
+     * @param id the ID of the game
+     * @private
+     */
+    private async waitForGame(id: number): Promise<GameSettings> {
+        return new Promise(async (resolve, reject) => {
+            const exitTimeout = 15000; // wait time is max 15s
+
+            const fetchInterval = setInterval(async () => {
+                try {
+                    // Check if the game is ready
+                    const waitResponse: string = await this.provider?.evaluateExpression(
+                        chessRealm,
+                        `IsGameReady(${id})` // TODO change when API is ready
+                    ) as string
+
+                    // Parse the response
+                    const gameSettings: GameSettings = JSON.parse(waitResponse)
+
```

Comment:
Should we add a `clearinterval` here?

Reply:
Good catch:

4ddbfa6a130074ada4eb492b559eb224644db479